src/gt4py/next/iterator/tracing.py

```diff
@@ -248,37 +251,82 @@ def closure(domain, stencil, output, inputs):
     )
 
 
-def _make_param_names(fun, args):
-    """Expand *args parameter with remaining args."""
-    args = [*args]
-    param_names = []
-    for p in inspect.signature(fun).parameters.values():
-        if (
-            p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD
-            or p.kind == inspect.Parameter.POSITIONAL_ONLY
-        ):
-            args.pop(0)
-            param_names.append(p.name)
-        elif p.kind == inspect.Parameter.VAR_POSITIONAL:
-            for i in range(len(args)):
-                param_names.append(f"_var{i}")
-        else:
-            raise RuntimeError("Illegal parameter kind")
-    return param_names
+def _contains_tuple_dtype_field(arg):
+    if isinstance(arg, tuple):
+        return any(_contains_tuple_dtype_field(el) for el in arg)
+    # TODO(tehrengruber): The LocatedField protocol does not have a dtype property and the
+    #  various implementations have different behaviour (some return e.g. `np.dtype("int32")`
+    #  other `np.int32`). We just ignore the error here and postpone fixing this to when
+    #  the new storages land (The implementation here works for LocatedFieldImpl).
+    return isinstance(arg, LocatedField) and (
+        arg.dtype.fields is not None or any(dim is None for dim in arg.__gt_dims__)
+    )
 
 
-def trace(fun, args):
+def _make_fencil_params(fun, args, *, use_arg_types: bool) -> list[Sym]:
+    params: list[Sym] = []
+    param_infos = list(inspect.signature(fun).parameters.values())
+
+    for i, arg in enumerate(args):
+        if i < len(param_infos):
+            param_info = param_infos[i]
+        else:
+            if param_info.kind != inspect.Parameter.VAR_POSITIONAL:
+                # the last parameter info might also be a keyword or variadic keyword argument, but
+                # they are not supported.
+                raise NotImplementedError(
+                    "Only `POSITIONAL_OR_KEYWORD` or `VAR_POSITIONAL` parameters are supported."
+                )
+            param_info = param_infos[-1]
```

Comment:
- This is a bit complicated to follow, but I don't have a proposal.
- What's the error if `len(args) < len(positional_params)`?
- Do you know where we actually need `VAR_POSITIONAL`? Probably in the on-the-fly fencil implementation and in some generic tests?